examples/hackweek_demo/plot_pnl.py

```diff
@@ -112,93 +113,134 @@ def get_wallet_from_onchain_trade_info(
+
+        # Replace positive deltas with share_price
+        share_price_on_increases = share_price_on_increases.where(deltas <= 0, share_price, axis=0)
+
+        # Fill forward to replace NaNs, to have updated share prices only on position increases
+        share_price_on_increases.fillna(method="ffill", inplace=True, axis=0)
+
+        # Calculate weighted average share price across all deltas
+        # weighted_average_share_price = (share_price_on_increases * deltas).cumsum(axis=0) / positions
+        weighted_average_share_price = pd.DataFrame(data=pd.NA, index=deltas.index, columns=deltas.columns)
+        weighted_average_share_price.iloc[0] = share_price_on_increases.iloc[0]
+        for row in deltas.index[1:]:
+            # index positive deltas
+            cols = deltas.loc[row, :] > 0
+
+            new_avg = []
+            if len(cols) > 0:
+                # calculate update
+                # new_avg = (old_amount * old_avg + delta_amount * delta_avg) / (old_amount + delta_amount)
+                new_avg = (
+                    share_price_on_increases.loc[row, cols] * deltas.loc[row, cols]
+                    + weighted_average_share_price.loc[row - 1, cols] * positions.loc[row - 1, cols]
+                ) / (deltas.loc[row, cols] + positions.loc[row - 1, cols])
+
+            # keep previous result where delta isn't positive, otherwise replace with new_avg
+            weighted_average_share_price.loc[row, :] = weighted_average_share_price.loc[row - 1, :].where(
+                ~cols, new_avg, axis=0
             )
 
+        agent_info[agent] = Info(
+            wallet=wallet,
+            positions=positions,
+            deltas=deltas,
+            open_share_price=share_price_on_increases,
+        )
+    return agent_info
+
+
+def calculate_pnl(pool_config, pool_info, checkpoint_info):
+    """Calculate pnl for all agents."""
+    #  pylint: disable=too-many-locals
+    session = postgres.initialize_session()
+    all_agent_info = get_agent_info(session)
+    position_duration = pool_config.positionDuration.iloc[0]
+
+    for _, agent_info in all_agent_info.items():
+        wallet, positions, deltas, open_share_price = agent_info
+        for block in agent_info.wallet.index:
+            position = positions.loc[block]
+            current_wallet = wallet.loc[block, :]
+            state = pool_info.loc[block]
+            if block in checkpoint_info.index:  # a checkpoint exists
+                current_checkpoint = checkpoint_info.loc[block]
+                maturity = current_checkpoint["timestamp"] + pd.Timedelta(seconds=position_duration)
+            else:
+                maturity = None
+
             spot_price = calculate_spot_price(
-                row.share_reserves,
-                row.bond_reserves,
-                row.lp_total_supply,
-                row.maturity_time,
-                row.block_timestamp,
+                state.shareReserves,
```

Comment:
There's a new calculate_spot_price that doesn't use lpTotalSupply in hyperdrive_pricing_model_sol.py which I think we should use.  It's a copy of what's used in the solidity.

Reply:
interesting, don't we need that to measure virtual reserves? I'll try and have a look.

Reply:
oh I think their bond_reserves variable in solidity includes the virtual reserves baked in
so solidity's
`spot_price = (initial_share_price * share_reserves / bond_reserves) ** time_stretch`
is the same as python's
`spot_price = ((initial_share_price * share_reserves) / (bond_reserves + lp_total_supply)) ** time_remaining_stretched`

Reply:
good catch